### PR TITLE
Refactor serialize_data with better param names and logic

### PR DESCRIPTION
**Pull request by bobleesj**

### CODE REVIEW

File: tests/test_serialization.py [hunk 1]
```diff
@@ -17,30 +17,29 @@ def test_load_multiple(tmp_path, datafile):
...
     generated_data = None
+
     for headerfile in tlm_list:
         # gather data using loadData
         hdata = loadData(headerfile, headers=True)
         data_table = loadData(headerfile)
 
         # check path extraction
-        generated_data = serialize_data(headerfile, hdata, data_table, dt_colnames=["r", "gr"], show_path=True)
+        generated_data = serialize_data(headerfile, hdata, data_table, dt_col_names=["r", "gr"], show_path=True)
```
Comment (bobleesj) [inline tests/test_serialization.py:27]: `df_col_names` and `df_colnames` are used across, so sticking with `dt_col_names`

File: src/diffpy/utils/parsers/serialization.py [hunk 2]
```diff
@@ -64,64 +64,69 @@ def serialize_data(
...
+    # Prioritize overwriting hdata entries with data_table column entries
+    col_counter = 0
+
+    # Get a list of column counts in each entry in data table
+    dt_col_counts = [len(row) for row in data_table]
+    dt_max_col_count = max(dt_col_counts)
+
+    if dt_col_names is not None:
+        dt_col_names_count = len(dt_col_names)
+        if dt_max_col_count < dt_col_names_count:  # assume numpy.loadtxt gives non-irregular array
```
Comment (bobleesj) [inline src/diffpy/utils/parsers/serialization.py:89]: Compare the number of `col_names` values provided as an input for this function against the number of col determined from the data table.

**Review by bobleesj — COMMENTED** [review]
@sbillinge ready for review. I was looking into this issue: https://github.com/diffpy/diffpy.utils/issues/27

Before addressing this issue, I had to understand the code and thought I could help refactor this.

File: src/diffpy/utils/parsers/serialization.py [hunk 3]
```diff
@@ -40,13 +40,13 @@ def serialize_data(
 
     Parameters
     ----------
-    filename
-        Name of the file whose data is being serialized.
+    filepath
```
Comment (bobleesj) [inline src/diffpy/utils/parsers/serialization.py:43]:
```python
    for headerfile in tlm_list:
        # gather data using loadData
        hdata = loadData(headerfile, headers=True)
        data_table = loadData(headerfile)

        # check path extraction
        generated_data = serialize_data(headerfile...)
```

Here `headerfile` is used for `serialize_data(filepath, ...)`. `headerfile` prints `/Users/macbook/Downloads/dev/diffpy.utils/tests/testdata/dbload/e1.gr`

**Comment by bobleesj** [comment]
> @sbillinge ready for review. I was looking into this issue: #27
> 
> Before addressing this issue, I had to understand the code and thought I could help refactor this.

Ready for review

File: src/diffpy/utils/parsers/serialization.py [hunk 4]
```diff
@@ -40,13 +40,13 @@ def serialize_data(
...
+        The file path whose data is being serialized.
     hdata: dict
-        File metadata (generally related to data table).
+        The file metadata (generally related to data table).
```
Comment (sbillinge) [inline src/diffpy/utils/parsers/serialization.py:46]: is there a reason we are not calling this variable `metadata`?  or `header_data`.   calling it `hdata` I am having to carry around too many pointers in my brain that I have to cognitively resolve as I read.

File: src/diffpy/utils/parsers/serialization.py [hunk 5]
```diff
@@ -40,13 +40,13 @@ def serialize_data(
...
     data_table: list or ndarray
-        Data table.
-    dt_colnames: list
+        The data table.
```
Comment (sbillinge) [inline src/diffpy/utils/parsers/serialization.py:48]: what is a data table?  A table with data on it?  I think this is the something like the column data read from the file or something.

**Review by sbillinge — COMMENTED** [review]
I started to review this, but @bobleesj like you I am a bit confused what is going on here right from the get-go.  The function is called `serialize_data` but it seems to be reading  data-file.   This seems closer to deserializing data, not serializing it.  We may want to get @Sparks29032 in on the convo here, I think he may have written it.

I think we probably want code that will dump to different file formats and read from different formats.  These are dumpers and loaders.  If we serialize and deserialize things I think of it as a kind of (safe) round trip.  the unsafe things is just to pickle and then unpickle but we need to be more careful now.  But sinnce we want people using diffraction objects, the serializers that most interesting to me would be ones that serialize and deserialize complete diffraction objects.

**Comment by bobleesj** [comment]
@Sparks29032 when you have time this Sat,  is the `serialize_data` method used in any of the diffpy packages? Wondering the motivation and how it's used in practice.

**Comment by Sparks29032** [comment]
I think this was originally added for the `pdfitc` rest-api project. I don't believe it is implemented in any packages.

These are dumpers and loaders, but called serialization/deserialization as we are moving from serial files to python-readable information.

This was coded a while back, so I don't remember all the details, but I believe these were coded specifically to dump and load PDF data (hence a data table [list type] + header information [dictionary type]) into serial file formats like `.json`.

Moving forward, it might be good to standardize so only dictionaries can be passed into these functions, and the dumpers and loaders can use specific file-format libraries given the extension of the file.

File: src/diffpy/utils/parsers/serialization.py (hunk 4 above)
Comment (Sparks29032) [reply src/diffpy/utils/parsers/serialization.py:46]: The hdata came from loadData's naming of hdata for header data information. Good to change to metadata.

File: src/diffpy/utils/parsers/serialization.py (hunk 5 above)
Comment (Sparks29032) [reply src/diffpy/utils/parsers/serialization.py:48]: These serialization functions were coded in tandem with loadData, so the inputs of these match outputs of loadData. This assumes files structured like .chi/.gr/etc. files.

**Comment by bobleesj** [comment]
Thanks @Sparks29032 

> I think this was originally added for the pdfitc rest-api project. I don't believe it is implemented in any packages.

To confirm, this function is used by `pdfitc`? Wasn't sure if the source code for pdffitc is publicly available. 

> load PDF data (hence a data table [list type] + header information

Yes, there are test cases for loading `.gr` files  and turning these into a single `.json` file.

@sbillinge As the next step, since we have an issues to write a DiffractionObject serializer/deserializer (https://github.com/diffpy/diffpy.utils/issues/195) in `3.7.0`, should we close the following issue https://github.com/diffpy/diffpy.utils/issues/27 that this PR is trying to slowly address?

**Comment by Sparks29032** [comment]
@bobleesj It is not used by pdfitc, don't worry. It was experimental.

**Comment by sbillinge** [comment]
We definitely want the dumpers and loaders and a pure serializer.  We can put it off to 3.7 though to keep momentum. The serializer will change the API though because I would like to give as an option a file path, or probably better a stream so we can have different backends, and have it load the DO from there. So the serializer would be nice now I guess?

**Comment by sbillinge** [comment]
Dumpers and loaders to group standards file formats as we have are a bit specific, but probably worth having since the large community is using these file formats

**Comment by bobleesj** [comment]
Since this is a PR, will copy and paste the comments by @sbillinge in the existing issue here https://github.com/diffpy/diffpy.utils/issues/195.

**Comment by bobleesj** [comment]
Closing - we will probably make a new PR while when we are dealing with `3.7.0`. For now, this PR isn't needed for `3.6.0` release.

**Pull request closed without merging**